lib/pbench/server/api/resources/database.py

```diff
@@ -0,0 +1,65 @@
+import sys
+from sqlalchemy import create_engine
+from sqlalchemy.orm import scoped_session, sessionmaker
+from sqlalchemy.ext.declarative import declarative_base
+from pbench.server import NoOptionError, NoSectionError
+from pbench.common.logger import get_pbench_logger
+
+
+""""
+A storage engine to save revoked tokens. if speed is the primary concern, redis is a good option. but if data 
+persistence is more important, postgres can be another good option too. Here I am using an in memory store to get 
+something working.
+"""
+blacklist = set()
```

Comment:
@dbutenhof this was the medium post I came across when designing this https://medium.com/devgorilla/how-to-log-out-when-using-jwt-a8c7823e8a6

Reply:
Makes me think that maybe JWT is the wrong technology here. In fact, I've found several "JWT is stupid for session management" posts that make a lot of sense; they're bigger and more unwieldy than simply generating a unique token and storing it in a DB (e.g. Redis) that we'll probably need to access anyway on most requests.

I suppose I'm really biased because I'm used to session token schemes, and the idea of a "session token" that can't be invalidated on logout just feels absurd to me. There's got to be a good validated Python package for managing session tokens.

Reply:
Not in favor of a global variable for this kind of thing.

Django has that, I had thought flask would have something like that.

Reply:
@dbutenhof @portante There are some popular solutions to work with sessions in flask as well like [flask-login](https://flask-login.readthedocs.io/en/latest/). My confusion is that session based authentication is based on the fact that user state is stored on the server’s memory and when user logs in, server then stores the session Id in a cookie on the user browser. However, by design we are writing APIs which are stateless, and a user can actually use the API via hitting the endpoints without actually using any browsers. Is it correct to define session authentication with RESTful APIs? 

`Flask-Login` is very handy in case we are rendering views from back-end. But we have JavaScript application served through the Flask server. So should we use Flask-Login or go with Dave's earlier suggested _`whitelisted`_ token based authorization option (as all our functionality in this case will be based on RESTful APIs). 

The `whitelist` option instead of `blacklist` takes care of invalidating the user token upon logout instead out keeping track of inactive/blacklisted tokens and would work similar to sessions. We can use redis to keep track of only active/whitelisted tokens and when user logs in then API returns the whitelisted token to the javascript front-end application which will store that token on the user browser in the form of some type of cookie but the lifetime of that token will be managed by the server.

Reply:
> @dbutenhof @portante There are some popular solutions to work with sessions in flask as well like [flask-login](https://flask-login.readthedocs.io/en/latest/). My confusion is that session based authentication is based on the fact that user state is stored on the server’s memory and when user logs in, server then stores the session Id in a cookie on the user browser. However, by design we are writing APIs which are stateless, and a user can actually use the API via hitting the endpoints without actually using any browsers. Is it correct to define session authentication with RESTful APIs?
> 
> `Flask-Login` is very handy in case we are rendering views from back-end. But we have JavaScript application served through the Flask server. So should we use Flask-Login or go with Dave's earlier suggested _`whitelisted`_ token based authorization option (as all our functionality in this case will be based on RESTful APIs).
> 
> The `whitelist` option instead of `blacklist` takes care of invalidating the user token upon logout instead out keeping track of inactive/blacklisted tokens and would work similar to sessions. We can use redis to keep track of only active/whitelisted tokens and when user logs in then API returns the whitelisted token to the javascript front-end application which will store that token on the user browser in the form of some type of cookie but the lifetime of that token will be managed by the server.

In one sense, the whole concept of "login" isn't stateless; in another sense, the actual login is a discrete event, generating an authentication token that passed statelessly to each subsequent request via an HTTP header to validate that a login had been done. This isn't really a difference between JWT and session management.

The browser cookie thing is really just one way for a browser to keep track of a session. I don't really like dealing with cookies, but some people like it because it's implicit. I really prefer explicit authentication headers. Any client can do that, including a simple `curl -H "Authentication: Bearer xxxxxx"` (another reason to avoid cookies, which are harder to manage here). The Javascript saves the token from login/registration in the browser storage and pulls it back out when necessary.

Reply:
@dbutenhof right so essentially jwt and sessions are same thing in this context where a client would send the token with authorization header that server validates or invalidates it. And how the client stores the token is upto the cleint

Reply:
So should I try to implement jwt tokens with simple [flask-sessions](https://flask-session.readthedocs.io/en/latest/) where a key is a user_id and value is jwt?